### PR TITLE
Insert shortcode

### PR DESCRIPTION
**Pull request by MakhonkoDenis**

### CODE REVIEW

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php [hunk 1]
```diff
@@ -0,0 +1,258 @@
+<?php
+/**
+ * Module Name: Insert Shortcode
+ * Description: The module allows you to add shortcodes editor tinyMCE.
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: The module allows you to add shortcodes **from** editor tinyMCE.

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+				// Register admin assets.
+				add_action( 'admin_enqueue_scripts', array( $this, 'register_assets' ), 0 );
+
+				// Load admin assets.
+				add_action( 'admin_enqueue_scripts', array( $this, 'enqueue_assets' ), 11 );
+			}
+		}
+
+		/**
+		 * Include libraries from the `includes/admin`.
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: Include libraries from the **`inc/`**.

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+					array(
+						'module_dir' => trailingslashit( dirname( __FILE__ ) ),
+						'in_screen'  => array( 'post' ),
+					)
+				);
+
+				$this->ui_elements = $this->core->init_module( 'cherry-ui-elements' );
+				$this->cherry_interface_builder = $this->core->init_module( 'cherry-interface-builder' );
+
+				// Include libraries from the `includes/admin`.
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: Include libraries from the **`inc/`**.

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+		/**
+		 * Register assets.
+		 *
+		 * @since 1.0.0
+		 * @access public
+		 * @return void
+		 */
+		public function register_assets() {
+			// Register stylesheets.
+			wp_register_style( 'cherry5-insert-shortcode', esc_url( Cherry_Core::base_url( 'assets/min/cherry-insert-shortcode.min.css', __FILE__ ) ), array(), '1.0.0', 'all' );
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: Maybe add a new property `$version` where to store a module version. 
After this current code converts to:

wp_register_script(  handle_name, script_url, depends, **$this->version**, true );

This solution to solve the problem of js/css files caching.

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+		 * Function add new shortcode.
+		 *
+		 * @since 1.0.0
+		 * @access public
+		 * @return array
+		 */
+		public function add_shortcode( $shotrcodes = array() ) {
+			$plugin_slug = $this->added_shortcodes['slug'];
+			$new_shortcodes = $this->added_shortcodes['shortcodes'];
+			$this->added_shortcodes['shortcodes'] = [];
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: `[]`-syntax started on [PHP 5.4](http://php.net/manual/en/language.types.array.php)

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-button.php [hunk 6]
```diff
@@ -0,0 +1,70 @@
+<?php
+/**
+ * Sets up the admin functionality for the plugin.
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-button.php:3]: UPD file header

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-button.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+		 * @since 1.0.0
+		 * @access public
+		 * @return void
+		 */
+		public function add_button() {
+			$args = apply_filters( 'cherry5-is__open-button', array(
+				'id'         => '',
+				'name'       => '',
+				'style'      => 'normal',
+				'content'    => '<span class="cherry5-is__icon dashicons dashicons-plus"></span>' . esc_html__( 'Cherry shortcodes', 'cherry' ),
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-button.php]: Correct text-domain - [cherry-framework](https://github.com/CherryFramework/cherry-framework/blob/master/cherry-core.php#L179)

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php [hunk 8]
```diff
@@ -0,0 +1,270 @@
+<?php
+/**
+ * Sets up the admin functionality for the plugin.
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php:3]: UPD file header

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+			if ( in_array( $screen->base, $this->args['in_screen'] ) ) {
+
+				$this->shortcode_list = apply_filters( 'cherry5-is__shortcode_list', array() );
+				$sidebar_list         = $this->get_sidebar_list( $this->shortcode_list );
+
+				$args = apply_filters( 'cherry5-is__insert-button', array(
+					'id'         => '',
+					'name'       => '',
+					'style'      => 'normal',
+					'content'    => esc_html__( 'insert shortcodes', 'cherry' ),
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Correct text-domain - [cherry-framework](https://github.com/CherryFramework/cherry-framework/blob/master/cherry-core.php#L179)

`insert shortcodes` => `insert shortcode`

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+		 * The function returns a HTML structure of registered shortcodes.
+		 *
+		 * @since 1.0.0
+		 * @access private
+		 * @return string
+		 */
+		private function get_sidebar_list( $plugins = array() ) {
+			$structure['shortcode-list'] = array(
+				'type'  => 'component-accordion',
+				'title' => esc_html__( 'Shortcode List', 'cherry' ),
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Correct text-domain - [cherry-framework](https://github.com/CherryFramework/cherry-framework/blob/master/cherry-core.php#L179)

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+					'id'          => $plugin_slug,
+					'type'        => 'settings',
+					'parent'      => 'shortcode-list',
+					'title'       => $icon . $title,
+					'description' => $description,
+				);
+
+				$output_html = '<ul class="cherry5-is__shortcode-list">';
+
+				foreach ( $plugin_value['shortcodes'] as $shortcode_slug => $shortcode_value ) {
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php:171]: Maybe code in line 166-172 put to the new class method.
And if replace code in line 167-169:

    $defaults = array(
        'icon'        => '',
        'title'       => '',
        'description' => '',
    );

    $shortcode_value = wp_parse_args( $shortcode_value, $defaults );

code looks like more clean ;)

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+			$shortcode_list = apply_filters( 'cherry5-is__shortcode_list', array() );
+
+			$plugin_slug = ( ! empty( $_GET['data']['plugin_slug'] ) )? $_GET['data']['plugin_slug'] : '' ;
+			$shortcode_slug = ( ! empty( $_GET['data']['shortcode_slug'] ) )? $_GET['data']['shortcode_slug'] : '' ;
+			$shortcode_attr = isset( $shortcode_list[ $plugin_slug ]['shortcodes'][ $shortcode_slug ] ) ? $shortcode_list[ $plugin_slug ]['shortcodes'][ $shortcode_slug ] : false;
+
+			if ( ! $shortcode_attr ) {
+				return array(
+					'error'   => true,
+					'message' => 'Shortcone not find.',
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: `Shortcode not found` :)

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+				}
+				$shortcode_option_template = apply_filters( 'cherry5-is__options-template', Cherry_Toolkit::get_file( $this->args['module_dir'] . 'inc/views/shortcode-options.php' ) );
+				$output_html               = sprintf( $shortcode_option_template, $plugin_slug, $shortcode_slug, $title, $description, $content_area, $shortcode_options_html );
+
+				return array(
+					'error'             => false,
+					'pluginSlug'        => $plugin_slug,
+					'shortcodeSlug'     => $shortcode_slug,
+					'twin'              => $twin,
+					'defaultContent'   => $default_content,
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Missing one space

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+		/**
+		 * The function returns content area HTML.
+		 *
+		 * @since 1.0.0
+		 * @access private
+		 * @return string
+		 */
+		private function get_shortcode_content_editor( $content = '', $plugin_slug = '', $shortcode_slug = '' ) {
+			$template    = apply_filters( 'cherry5-is__content-area-template', Cherry_Toolkit::get_file( $this->args['module_dir'] . 'inc/views/shortcode-content-area.php' ) );
+			$title       = apply_filters( 'cherry5-is__content-title', esc_html__( 'Shortcode content.', 'cherry' ) );
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Correct text-domain - [cherry-framework](https://github.com/CherryFramework/cherry-framework/blob/master/cherry-core.php#L179)

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+
+		/**
+		 * The function returns empty layer HTML.
+		 *
+		 * @since 1.0.0
+		 * @access private
+		 * @return string
+		 */
+		private function get_empty_layer() {
+			$text     = apply_filters( 'cherry5-is__empty_layer-text', esc_html__( 'Shortcode not a have options.', 'cherry' ) );
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Correct text-domain - [cherry-framework](https://github.com/CherryFramework/cherry-framework/blob/master/cherry-core.php#L179)

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+				'description' => '',
+				'icon'        => '',
+				'slug'        => '',
+				'shortcodes'  => array(
+					array(
+						'title'       => '',
+						'description' => '',
+						'icon'        => '',
+						'slug'        => '',
+						'twin'        => false,
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Maybe `twin` argument replace on `enclosing` or `selfclosing`

- https://codex.wordpress.org/Shortcode_API#Enclosing_vs_self-closing_shortcodes
- https://generatewp.com/shortcodes/

File: modules/cherry5-insert-shortcode/inc/views/insert-shortcode-pop-up.php [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+// If this file is called directly, abort.
+if ( ! defined( 'WPINC' ) ) {
+	die;
+}
+?>
+<div class="cherry5-is cherry5-is__modal-window cherry-ui-kit">
+	<div class="cherry5-is__popup">
+		<div class="cherry5-is__popup-header">
+			<div class="cherry5-is__popup-header-inner">
+				<h3 class="cherry5-is__popup-title"><?php esc_html_e( 'Insert Shortcode', 'cherry' ); ?></h3>
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/inc/views/insert-shortcode-pop-up.php]: Correct text-domain - [cherry-framework](https://github.com/CherryFramework/cherry-framework/blob/master/cherry-core.php#L179)

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+		 */
+		public function enqueue_assets() {
+			$screen = get_current_screen();
+
+			if ( in_array( $screen->base, $this->args['in_screen'] ) ) {
+				wp_enqueue_style( 'cherry5-insert-shortcode' );
+				wp_enqueue_script( 'cherry5-insert-shortcode-js' );
+
+				$dev_mode = ( constant( 'WP_DEBUG' ) ) ? 'true' : 'false' ;
+				wp_localize_script( 'cherry-js-core', 'cherry5InsertShortcode', [ 'devMode' => $dev_mode ] );
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: `[]`-syntax started on [PHP 5.4](http://php.net/manual/en/language.types.array.php)

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+		}
+
+		/**
+		 * Function add new shortcode.
+		 *
+		 * @since 1.0.0
+		 * @access public
+		 * @return array
+		 */
+		public function add_shortcode( $shotrcodes = array() ) {
```
Comment (cheh) [inline modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: `$shotrcodes` => `$shortcodes`

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php (hunk 1 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: Done

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php (hunk 3 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: Done

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php (hunk 2 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: Done

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php (hunk 18 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: Done

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php (hunk 19 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: Done

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php (hunk 5 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: Done

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-button.php (hunk 7 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-button.php]: Done

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php (hunk 9 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Done

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php (hunk 10 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Done

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php (hunk 16 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Changed on `enclosing`

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php (hunk 11 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php:171]: Done

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php (hunk 12 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Done

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php (hunk 13 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Done

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php (hunk 14 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Done

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php (hunk 15 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php]: Done

File: modules/cherry5-insert-shortcode/inc/views/insert-shortcode-pop-up.php (hunk 17 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/inc/views/insert-shortcode-pop-up.php]: Done

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-button.php (hunk 6 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-button.php:3]: Done

File: modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php (hunk 8 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/inc/class-cherry5-insertion-popup.php:3]: Done

File: modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php (hunk 4 above)
Comment (MakhonkoDenis) [reply modules/cherry5-insert-shortcode/cherry5-insert-shortcode.php]: Done

**Pull request merged**